backstory/src/test/java/com/google/sps/story/DatamuseRequestClientTest.java

```diff
@@ -0,0 +1,339 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.story;
+
+import static org.mockito.Mockito.*;
+
+import au.com.origma.perspectiveapi.v1alpha1.models.AttributeType;
+import com.google.gson.Gson;
+import com.google.sps.APINotAvailableException;
+import com.google.sps.story.data.HttpUrlStreamHandler;
+import com.google.sps.story.data.DatamuseRequestClient;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.net.MalformedURLException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.net.URLStreamHandlerFactory;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.ArgumentMatchers;
+import org.mockito.BDDMockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/** Quality tests for DatamuseRequestClient */
+@RunWith(JUnit4.class)
+public final class DatamuseRequestClientTest {
+
+  /** a default url to test with */
+  private static final String DEFAULT_URL = "http://sample-website.com/fake?";
+
+  /** the stream handlet to use for tests in this class (necessary to mock URLConnection) */
+  private static HttpUrlStreamHandler httpUrlStreamHandler;
```

Comment:
Can this not be static, so that state is less likely to carry over from test to test? (having a clean slate for each test avoids the need to wonder whether these tests are order-dependent)

Reply:
As you noticed below, it does have a clean slate for every test. Doing this saves the effort of instantiating this class before every single test, when all you need to do is reset the connections. I think just resetting connections efficiently accomplishes this. However, if you want me to change it, I'm happy to. @ykjchen

Reply:
Yes, don't worry about saving this level of computation during tests. Confidence in correctness is much more important.

Reply:
Ok I'll change it.

Reply:
Done